### ISSUE TITLE
Parametrized stroke outlining / expanding / offsetting 

### ISSUE THREAD

**Issue by strandedcity** [paperjs/paper.js#371]
I've been following the library for some time, and have started a spare-time project based on it. I've been eagerly interested in the functionality to expand strokes into paths, particularly for bezier curves (where I know the problem is mathematically complex). Has a strategy / timeline for implementation been discussed? I may be able to contribute if new contributors are welcome.

I wasn't sure if the "outline" feature was separate from what I would describe in a CAD world as "offset". Is there some way to get equations for the position of the stroke edge that is not as complex as a full bezier-offset solution? See http://pomax.github.io/bezierinfo/#offsetting for a very detailed explanation of the math involved.

Thanks!

**Comment by lehni** [comment]
@iconexperience I haven't seen your plunker before, it looks really good! If I understand correctly, the main issue that's stopping you is in the boolean operations code, correct?

**Comment by iconexperience** [comment]
@lehni Yep, a fully working boolean code would solve almost all the remaining issues. There are some edge cases that would need a little work elsewhere, but a boolean code that really works would be a dream come true.
I am extremely busy at the moment (a few days before new product release), but if there is something I should test I will try to arrange a little free time. In a few weeks I should be able to contribute again to PaperJS.

**Comment by lehni** [comment]
@iconexperience that's great to hear. There are still real issues remaining in the boolean code unfortunately, mainly #449 and #450. Unfortunately I haven't heard from @hkrish in quite a while, not sure if he's still working on this other approach or not. We could attempt fixing this in the current code-base together? Your code in #648 would probably be of use here?

**Comment by lehni** [comment]
@kuribas I finally wanted to look at your outline of the boolean code posted in https://github.com/paperjs/paper.js/issues/371#issuecomment-59438784, but that page is gone. Has it moved elsewhere? Would you be interested in joining the effort of implementing proper boolean operation code, as well as offsetting in paper.js?

**Comment by kuribas** [comment]
I am currently debugging my sweep line algorithm for set operations on paths.  It's written in literate haskell, and hopefully clear enough to understand the workings.  I'll upload the file when I am finished.  

As for offsetting, I have found an algorithm to quickly find the best parametrization for a least squares fit.  I then use a binary search to find the best nodes to subdivide.  It is still rather slow, so probably not very useful for real-time applications.  In my tests a least squares solution with a simple approximate parametrization and subdividing the curve will give good enough results (which I believe is what you are doing now?)

My javascript knowledge is limited, but if you need help understanding my algorithm, I am glad to provide it :-)

**Comment by kuribas** [comment]
For finding intersections I use the bezier clip algorithm (http://tom.cs.byu.edu/~557/text/cagd.pdf).
My implementation is here: https://github.com/kuribas/cubicbezier/blob/master/Geom2D/CubicBezier/Intersection.hs  
An alternative (faster) way is implicitization.  It would reduce intersections to finding the root of a maximum 9th degree polynomial, which can be solved by http://ir.nmu.org.ua/bitstream/handle/123456789/111417/40ffff39f2dd54b3467a4431e412d181.pdf?sequence=1
I am considering using that if the intersections are a bottleneck.

**Comment by kuribas** [comment]
That last paper is also interesting since it contains the closed form formulas for finding roots of 3rd and 4th degree polynomials, which is quite a bit faster.  For example finding the intersection of a line and a bezier curve reduces to finding the root of a 3th degree polynomial, which I am using in many places.

**Comment by microbians** [comment]
Sorry to insist, but for offseting and squares fit I think could be useful read my paper ;) 

http://microbians.com/?page=math&id=math-quadraticbezieroffseting

I know my english is terrible, but I think is a very fast algorithm, with a loop of 3 subdivisions the offsetting  is more than usable and very fast, here is a guy that do a html5 test using my algorithm:

http://brunoimbrizi.com/unbox/2015/03/offset-curve/

He do it using only one interaction of the subdivision, but if you do two you can arrange a very good aprox. 

![captura de pantalla 2015-08-19 a las 10 19 31](https://cloud.githubusercontent.com/assets/1662136/9351672/cfe8b48c-465b-11e5-98fb-7b206df797b7.png)

My flash example (sorry is flash and old) http://microbians.com/?page=code&id=code-bezieroffsetingplayground

But the thing with bezier clip, boolean op, and that is another story ;)

**Comment by lehni** [comment]
@microbians your approach looks great, but works for quadratic curves. Since we're dealing with cubic curves in paper.js, I have a feeling we need a different approach. We could of course convert cubics to quadratics first, but that's not a trivial problem either.

**Comment by lehni** [comment]
@kuribas Unfortunately I'm not fluent at all in Haskel, so we'll have some language barriers both ways. I am looking for a fast approach indeed. Did you see @iconexperience proposal? It appears to produce very nice results. I'm curious to hear your thoughts on the math behind it.

**Comment by lehni** [comment]
@iconexperience I've been playing with our code quite a bit now and am impressed so far with how it behaves. I'd be curious to find out why your approach produces good peaks, and how you go there : )

"The peaks are calculated by finding the roots of the dot product of the first and second derivative."

**Comment by lehni** [comment]
@iconexperience regarding the merging of the outlines of the separate curves to full paths, as outlined here https://github.com/paperjs/paper.js/issues/371#issuecomment-59381879:

Due to requirements in hit-testing, there is code already that can create bevel join and square cap geometries, see Path._addBevelJoin() and Path._addSquareCap(). I've designed these with path offsetting in mind, so they will come in handy when producing the paths to be used for the boolean ops. The only part missing is round joins, but that can be handled with out boolean operations, simply by drawing a connecting arc. I have this all pretty clearly mapped out, so once the boolean operations can handle touching paths, we should be ready to go!

**Comment by kuribas** [comment]
The math looks fine to me.  The offset curve however can also have singularities at different points as the curve itself.

**Comment by kuribas** [comment]
It would be nice to see the difference in number of curves between your algorithm and a simple midpoint algorithm (for a given tolerance).

**Comment by lehni** [comment]
@iconexperience while I am working on fixing the boolean operation issues, I had and idea for how to handle path offsetting along stroke outlining in a way that would not even require the operations here. Let's work on both, so v1.0 can finally roll out!

**Comment by iconexperience** [comment]
@lehni
I have updates my plunker to use the current boolean-fix branch. The algorithm for the offset curve has been replaced by a much simpler one that does not create those nasty little self-intersections that we saw before. Also, panning has been improved to respect the zoom factor. Here is the new plunker:

http://embed.plnkr.co/jkNGoe8fD0li9DmY7u76/preview

The results are impressive, but there are still many cases that do not work. But as far as I can see, all these cases involve self-intersecting paths. On the other hand, there are already many cases with self-intersections that work.

The general problem with the new curve offset algorithm is that it can result in offset paths like this:

![image](https://cloud.githubusercontent.com/assets/2122773/9597412/a3baaade-5081-11e5-99f8-b5c1ad8180d4.png)

This has nothing to do with our boolean code. It is not nice, but not really a tragedy. Also, I have seen this problem in professional imaging software, so maybe there is no trivial way to solve it.

Anyway, this new algorithm makes it much easier to get good results. And I have the feeling that we are not too far away from a working version.

**Comment by lehni** [comment]
@iconexperience sounds very promising! Does your code life in a git repository? Would be nice to see the changes you've made, and perhaps contribute some too! I've copied the previous version locally and saw some possible code simplifications that I'd like to contribute back.

Also, for merging the sub-paths I think I have an idea that doesn't even require boolean operations. I'd like to give this a go soon, but only once your code has stabilized I guess.

**Comment by lehni** [comment]
As for the strange path, the previous code created a self intersection there, correct? Wouldn't it be better to allow these then, and resolve them filter the "islands" out?

**Comment by lehni** [comment]
What I mean is: I wouldn't necessarily try to avoid this overlap-triangle here. I would try to detect it and cut it out, because the overall outlines look cleaner and more correct than in the new code.

<img width="419" alt="screen shot 2015-09-01 at 10 28 40" src="https://cloud.githubusercontent.com/assets/213275/9599438/3bf72b6c-5094-11e5-9e91-86325c9b79a6.png">

<img width="372" alt="screen shot 2015-09-01 at 10 29 10" src="https://cloud.githubusercontent.com/assets/213275/9599447/52151bfc-5094-11e5-92c4-0dec3ceb35e4.png">

**Comment by lehni** [comment]
Also, since you're using the `cleanUp()` function which resolves self-intersection, all the failing boolean ops that are left are actual failures of the current code base, and not due to self-intersection. Would be good to extract some and create new issues for those!

**Comment by iconexperience** [comment]
@lehni 
Sorry, no git repository yet. I will try to do this.
I have tried another modification to the code, which completely eliminated self intersections in the segment offsets. This improved the result again, but there were still many cases where the boolean operation failed.

Therefore I have reverted the code to the original cuve offset algorithm, which will create more issues. If the boolean operations work with this, they should have no problems with the simpler curve offset algorithm.

I think we should decide later, which algorithm to use. Let's now focus on the boolean code. I will open an new issue with some examples later today.

**Comment by kuribas** [comment]
The problem of self intersecting curves can be simply resolved by splitting the curve at vertical tangents (which is required for my algorithm).  That's as simple as solving a second degree equation.

**Comment by nicholaswmin** [comment]
The *variable-stroke-expansion* that @microbians shows (here)[http://microbians.com/?page=code&id=code-theelectronicsketchbook] is really slick.

Is this stroke expansion via geometric path-offsetting or is it a stroke rendering trick?

Do the approaches outlines above cover such functionality? It could work wonders for Pencil Tools and sketch-like drawings.

**Comment by microbians** [comment]
Its a geometric interpolation ;) in reality it offset the quadratic bezier for the minimum and the maximun and interpolate the data in relation to ( length of the quadratic bezier segment to the subdivision point / the length of the quadratic bezier )

**Comment by lehni** [comment]
@nicholaswmin yes I am planning to support such a feature once we have constant offsetting working. Since we're dealing with cubic bezier instead of quadratic ones in Paper.js, it'll all be a wee bit more tricky, but the interpolation trick mentioned by @microbians should work similarly.

**Comment by nicholaswmin** [comment]
Should I open a new Issue for this so it can be tracked? - I'd hate to pollute the path offsetting Issue with discussions about this - This is something that other Canvas libs simply don't support or even mention, but would it be something worth having in mind for now? I'd love to see a feature like this in Paper.js.

**Comment by frankrolf** [comment]
I don’t know if this is of interest, but checkOutlinesUFO (https://github.com/adobe-type-tools/afdko/blob/master/FDK/Tools/SharedData/FDKScripts/CheckOutlinesUFO.py) – which is used to remove overlaps in fonts – also uses pyClipper; inspired by Robofont. 
The flattening of outlines only happens as an intermediate step, the final outlines are of course not flattened.

**Comment by benoitvb** [comment]
@iconexperience Is there a simple way to make it works for straight path as well ? For now, it brokes as my path does not contain any curve. Thanks for the help !

**Comment by kuribas** [comment]
Flattening is less efficient than bezier clipping, especially when high precision is required.

**Comment by lehni** [comment]
@iconexperience I've found some information regarding your dot product of the first and second derivative (what you called "peaks"): 

http://math.stackexchange.com/questions/1954845/bezier-curvature-extrema

It looks like this simple calculation will give you something that resembles curvature extrema, but isn't quite those. I can confirm this in my tests. But finding the actual extrema is more expensive, we'd need a solver for quintic polynomials... I am wondering what we should call these peaks?

And while digging deep inside Skia's code, I found out that they do wrongly call those curvature extrema, and base their code on this assumption:

https://github.com/google/skia/blob/master/src/core/SkGeometry.cpp#L784

It's funny that this hasn't cause any issues yet. : )

**Comment by lehni** [comment]
Some more information about curvature extrema VS "peaks". I created a quick demo that compares the two methods on an interactive curve:

[Demo](http://sketch.paperjs.org/#S/zVhLc9pIEP4rUxxiSSBAQoBDHhdvDqnaJN5Kai+Yw2gYQLYsOXpgqVL89+3uGYFG4N3aVG1tdMCjfnz9nB6Nf/QS/ih7i97XB1mIXW/QE+ka36NExOVaWle7onjKF6ORWCfDjD9vo2Io0sdRLHdJNJq+DjzP92Qgg3C9ngbTYCzn08AfS2885bNwBBqj+WQjpNjMAm8uZ6E/9aU3n8nXk7Hnb/jE866v+UyMr+XoKY1rN0vTIh/e51f2m7vkLtnzjMVRItk7trxLGDzLpTceD5g/Hq9Wg4bkH0l3yeqo+JRGSfEPmt5FTVHmTxcVBywp43jAFMX1LnsBfLcl0EbOo2Qbn4fjtbGnQHgR158DbDAzUXkmdj+PiY5OlZ/H1PECARP5zG5haf1QerncPsqkyBeUIQ2WF1n6IG/SOM0W7CqMuXi40qwNGK+/yliKQq4XrMhKeZccTqXdRcVxXaRprE1+g6WlpJA6TJNPaZnL39LnBCQ2ZSKKKE0suQdfbKZ9AyxgouNDWH6TeaEEhtQGA/YDLMQy44mQC8jNoBUM+jVgO56sY6lfGXmJuHG6JV8OZ+5kfPuyOxgSbi6Q0BR8rrTRK8gU+dVkipjKAzdKkK1ePiaXJNKyOIl8gRclclhiEl69wlwMi/pJrnQM0YZZ6Izd9gWFtDtLZK5Y/x1TKVvLuOBa9/BSGjiEveeFJHLDvkuafCiNdjpEme1lU6NNlOXFDVJa8e15XMocREh0uJXFn0SxmmIABXV4UWbyQ1Vk8pFbSskebtLsAxc761iQwgi36eXhTZSJWFotFj4C4pbQwDeN5VssjwYfsMIemPIZX0clNMu0Qzd3wzaT0qjgwR6C0+lefkm6Hqyhn5peXEOrN/vlpNpUpJWNW8kf8v85BZO/TUEm1/9JAg5Gs4G3H5MNDhp4g4QYjVd50FL7JQ67Wi29Vculyieaj2y1nBjsCdECZKvl1GAHRJshWy3nbTavgAYWXPCiRQ6JPEEy8njV4gniBcibEA9+fOaAThu3BqEadWsDl8ioRjxet3GJh7gkAAgaty1EBzCeJc3s+Fw+yiwSPB7mabyXf5Qc6lREwjKLBwE5aMFFF2BVdZqCH/n8ZX545IddPvmFdR+NzshwtMFx13RHJmE6JIpzqU/0lvm1OsQFD/psApFTP9BiApQq6DYNsTyQmSlhrTXpNpCrBRvUbu+4dcOrG4s1WqyDbjsRq7FYNxbrSbe1XC3YoP6Lrropw7OOeg0gFrUF/PRVW/C6O4ROUiFKhZelsM2t0JQKa1u7aomjFXFZX0mcdMWZxE804vlJ9sv0pGg6zdKjy77UhpYeYDb8GhuanyaYRzwcC1W3XwhCjzD7UtNZepARfn02/GiSEb4aO4YDmBJeOdycmUStHWMshiQbVo7hYUiyYe0YuIJkReUIMxqi1o7pZEWJ4h1gXlN8vIPMa8o5kDuZpFQBiDB9JhB0umNSS3ewQ41telL5OiM0RrixYyErOgXEDA0mXo58b+YoBMciuy7ZaRfyOw4F73rsYCFqxwqpmUIsaJ8h3eLq9AOKQyH1mT8fH2FdSkqfktDGzdB84CgXTevwdw6kQJmkDWup8KaOMlPbDoG2xdCwNXWOeUBvbDPjOL1ARUfiChIVqjc9gJo4lF6VK8fyMAoy703Igq3NoiyUzUKyFhgrARxGDpUOQWZGckTXHbzruME1YmDHoAa+QcSOoC11yo9ASMxsQJZ1YeGdMlepzNUXbJRgY4bgEK/f1AdLYpN/AnNv0OljBUqFRwWwEZ98c5nfLj5aunA2CPwHgLV8GrDvA5bB/Qw+OQesXNkwAt+cxmIm1UHSQsBLxelA+XD79ePvXz5rFfgkZhbqRSA0fgN/3iqLw1gm22IHlH7f+EqmyygIk9Qyamzjg/eogr0HIAhF4k2rYG/fMTzyzLtVU6JP+LX9GCWWWvDKgpDGduuQaFtNH0AFv3k7vGMM9yoGkAPT9xiJPMVx34mjeQiV7PMwB/dd1Frer2z2nnnSnXWMHcxXDDl9uAiMxp/KfAe3jJcxDsYtsjkSJR2IvUEvzOBrjG7CeW+xXB3+Ag==)

- Red circles are "peaks" (dot product of the first and second derivative)
- Green circles are curvature extrema (found by solving a quintic polynomial through the external poly-roots.js library for now)

This may help figuring out further which information is more useful for sub-dividing curves before performing offsetting. Perhaps we do need an internal solution to finding the actual curvature extrema?

**Comment by vcasagrand** [comment]
I've worked on a script to generate stroke outlines from a shape with line segments (non-bezier) using miter joint, I think it can be improved, but here's:

[Sketch](http://sketch.paperjs.org/#S/jVfbbuM2EP0Vwg9ruStrLW2DAg7aYpFekEW3LRoD+2D7gZbGthCaNCTamzTwP/UTO8OLRPkSx0Aiau48MxyOXnqSb6A37j08gs7XvbiXq4LelzuZ61JJllfANfzN9ToasJeZZPjb84ptkcJ+ZBK+McN0LPrVsNqA1PWYTac3o5jdjOYxm6YjXP5glwGVlmkgkX40a0+ex63dXKgaijHT1Q4C8nInxPMDCMi151rmYXBrFxRrUutKPcLXsjBxp5njVaB3lTQiM3mYyZlstm41/tppUUqISg2bmG3wUf0BcqXXBo6ZJjAqqHdCo9klFzXcerJU1YaLGA3xSqfumcUMZJGa/7jmMVvgf1wt8K+s75TcwxOuJDqqwUTSGFTLZQ3kh4Lp7Og7NkpuGjnBa323q/bgRXN6qafBOhFmE2zI0nlH708TNCo2RpIVOOonHY0GHekJlyvM9bG4I5/K/4pbd8LWZLJB5MqteI6GdneDhBdF1FrbKoQiO7GSXbDyNiOrSu22dRedZMO3kU9+ZGguxQ58Wwiu6Ld8C1Viz8WtFXICLi93SqgKhfsLsYP+WRFfjFSLli89+Pkl4GdaN5DnF+GeaVtzFOwrAOUtOGlHMTujOLym6TdHfOv+IifzHLD18FZfod65GK+phYGQ6wv0s/LdLbV1gdgLqorI9EJyF7XNqUlX8oRn9As1olzV0TA4O4kxcC//4UXJZT3AI+l1nr1OXcpXdeJXPV7TZu9PPV6LsnU4GFheU8QfPrDPBEPdYmTaZnr56Fj+8clZVQDSH53GTHbFTHZspoLCGzGBPpQriXSbt5/YyPM8zfB/dl2bjV3bdlKLI6mmHY2b/ubtUaDclI5Ht63v8B7x0S9IYdF0rsmx0jkdBxzp8HO07AgaIi7O0RrB8OLBeJwxbDL3AaO2wWSD6Wju9ZYsCnWb3un8nGZtyrvXnD2r5j4M2wvejnMf3IEBXrBvsuyv1NesBTC81rVDgU7P7vQBDU/aT0TkaoLv4VhkJhFijNvGbQQ/6Um5gQgv8PAcm4lHITykYCou0eq38gkKbPJ4Yvv/9Y+kl6UQJv4xhc/zxxMBNPcVytVak4QSxTmBh/JfGLP0+5Zj5ii7z3CwMJu4benN4GCbcsBphwR3EAKeGwigPTfBBU+tuPHtJjXC93e6vqOp685uKkvdM3PP2GTEJdvu4dSGHQSIaWY/yuO6pPhMRpUSLqMTXEYOB0n0RMkvalfDL+obHZNmboA97q+ZlNEWMs38icsJ1NoK+FJ/QR8CKi5zwhxn3nZ2NlMuW3NZCHCvNhOHkxAqvrocgr/CKtioPUR+JjbzKc7+qBjUaN/572OBmBDDCunbYIalJLZ9uZfnJNBnK4IDdN8N5FPC4907giXRz1uYu2Cod1AwgzAWEnLhTIk5Z++xIgx6BQjNne6hs03cTndsJ+xj9nHURY4jUnv8rjFwBJ8z4deO41w3jF9NC9R7NIDVvfF0fvgf)

**Comment by microbians** [comment]
Interesting doc related to WebGL + Quadratic Beziers
http://wdobbie.com/post/gpu-text-rendering-with-vector-textures/

**Comment by lehni** [comment]
Looking for further resources on Hoschek’s method, I accidentally came across this really curious document that describes another method for curve offsetting that claims to be much faster and as precise, if not better:

"A New Shape Control and Classification for Cubic Bézier Curves" by Shi-Nine Yang and Ming-Liang Huang: http://www.springer.com/de/book/9784431684589?wt_mc=ThirdParty.SpringerLink.3.EPR653.About_eBook

Here the chapter: http://link.springer.com/chapter/10.1007%2F978-4-431-68456-5_17

And here the Google Books link: https://books.google.ch/books?id=HJeqCAAAQBAJ&lpg=PA204&ots=YwBnF5ffia&dq=%22A%20New%20Shape%20Control%20and%20Classification%20for%20Cubic%20B%C3%A9zier%20Curves%22&pg=PA204#v=onepage&q=%22A%20New%20Shape%20Control%20and%20Classification%20for%20Cubic%20B%C3%A9zier%20Curves%22&f=false

I have started experimenting with this recently and the results are very promising. I will post my findings here shortly.

**Comment by lehni** [comment]
I've extracted the pages from google books and turned them into a PDF for easier reading :)

https://www.dropbox.com/s/0usbx7e65xz5emb/A%20New%20Shape%20Control%20and%20Classification%20for%20Cubic%20Bezier%20Curves.pdf?dl=0

**Comment by lehni** [comment]
And here a [first sketch](http://sketch.paperjs.org/#S/fVVNj9owEP0rFhfCNqQBdtuKVQ7V9tAeqlZa1EvgYJIBUoyDEoclWvHfO+MPcGC3PqDxvOf5eOOY157kO+hNe89bUNmmF/ayMqf9gVdsz9WGJUzCC/uNZvA6lwxXDesdSFVPWWoctNJ0FMchG8fxImSyESJkxjPE38Ui9Jn3F2Y69EiG47i1qsotPJWirKasvxQ82/YttML47TMIyBTkU6aqBubyNHicy7mkwrOiygR4pUdP2uM62JeFVFg+Zo5duornRYM9PbgchRAueQV5v5NgU6izrcpS2FQzNAPDIm9Uyp9lU8O38kUiY9XITBWlDOCA8g2YLQZjIbivyr/YToS7GdTKcCJdaOiYtFQpoOIygykbxZ6ql6GQGh6w4TIXcOun/px0tNfdnW5Kr/j6/dKLFQuw4IFfoPVFhYIdS5LEDqPDoWXc2GFdUGT2IWGm5xyE4o+MdekVZFxkjeAKtMLOf2IgargOTnOhi42lXyG0+lasPo5WK9wP3yAZ4YaFJJrZ/JD/Y5aNulB/4aZLPaUki2r3sHjsIiQZVXujkZ4fHrL1pkRaXCt1laW7bfb5jWJ23v60OU734Ijembl0kz9HcjVezy/Rz0W0KqpaPTXVAaI1qN+k7lc1K3YQxNHD+Yqdo3bG6kLrT5hC3Ab15D8gmrlEf7hooA4G/hfBkusiPfQYI3xI6Q1qjTnyH6njSPvGBBtz0oHH2ndPsDEfOvBE+z4RbMzPHZhKq6MjosZqPVDFfl8zLtdwkdDvT03eJ4584poiqpjybazpJ1xTHDXRsDV9OAd6ngIMckf4kGLc4aEB/k583hFpX9CJ3Q3ZPRmxMzq81vJax2sdr/V5nLJiUJsVj5msH6kij7e88OIzL7Y8e+2NSjWo7/rbHAWKeuCDt+FxgNLeseFSX1f8S1xWwLf6oah703Rx+gc=) that allows playing with this mentioned shape control method by dragging the red circle and all the curve handles.

**Comment by lehni** [comment]
[And here finally the offsetting sketch](http://sketch.paperjs.org/#S/pVdRj9o4EP4rFi+EbsgSrj2doFTq0YdWanuVdnWnE8uDSQy4a+zUduhyFT/4/sWN7ThxQlhVOqTdOJ7xzPibmc/OjwHHBzKYDe4eic72g3iQidy8396it0UhBc72SGzR7//+Q4lEWSmPBN63imjKd2iDFcmR4LMHfntr/hBCD4O36DP5ju72uCBoKbiWgiHMc7RkWCm6pRnWVHC0FRItyw3NvPWlsa4eBpWhuz0df6acoL8xuDIGPoHP8UdqXt+X8N95feBHLKug0AL9Mpm7GSKlkPd7SdResBwkk2T6au71C6z3MMch0i8wjH48cPCJFNkdCNdqhlZuwvxWq3QyidF0MlnHiJeMxcjNjOH/eh2Hmi8bzdU4UHI6XlcBKI9kKZiQMzTcMJw9DivRFuyf7ggjmSb5DGlZkgd+HtWBuxws7AaSLZVKW9zmbSC+uO1ZpYwJTqJRAtN+m7uS5sQZ73e7xUx5v43VZeW78dGNYFvyzGY3c4m3ksjGHKNCUK5HqIrBZiEFa1aaWGEaYFlM27JpINPNuh3R91ALkLS3+p4eSDQZhYrT64ppqJjb4tFpkkmhVKSnI/QC/YJu0ctQyUBqYkkOJdO0YKdoCriWGy1xpqMiTXCeR8V0FFrGsCgPzN4aX4F8E8rTSj53chc5YP0e6p+RFBQa13g06lebgp9GbbyxeudWenZVMj9Cg33gmkhFrEBFxwwS9QR/pxgd4Xk81RmDtlxilpUMawINuWMEaYH0nqCnMX6iCkVpjCajpMmvU1qgT6ZUsMZ8Go2P1nCIkKLc68AwsotCeSaUl8PwUg5xRZI9gXPJToDgAkUTE0hLo4ncBOy66IhZSZQhODMnhQZxXqHZLJbHDEyuwj6XQmhlJ6sMAJlFZsPU8MwcHq/Rb+Zxs0DTGj+PyhMoHbMVXaOxgboRmt+pEqIblFqF07wVSlKUah+11zxBrRqQxrD6hYEzvpQbkG+sHDR94ZzrxN4JBoBQbWn5ZNkS/UWAHwEQSDLlGStzwM5IYpRaPmbQM1DwH4Vj9GhUW8tdWWwpg9IC1o6NDdCC04QeCiGhEjQi+Q4SAUeIqsBeujI2kdiDIZKmGE1NxQ5xO0p98JLoUnIn8QUelHhZ5JDOqEZ/b3vDklYhFLU6ATd8MX3tKSR5ddkykmS+hBqjfTTX48iHHNBoRXgd/oMU1RF9FvKAWc1qkDm3vMeWc/ksKV6LoMuyVyNIfyICY2x8lW7nPZgFNuLrybgOixH6sLwDe0yKkptiTX8NeuFYn11Lb+1PSwFRSCf24mD72M3Zd5vOI1V0YwnNno9N76O6+VPX/NY9DG9uLrrfREWB5a1Kp1GLaxWpRx3NVu22ILlQ9WT1DOcD3SeG8xPgZm5GPDl1zeRUmdCn3SpoQfDV8QODh3WbMMJ3ej8HwWvE4NEBxIPy7TIz4fatrdXXdTcmG5c5lM2CdxAg5hmJvo3mbTW6RVEOAZgtjHosuJ3lwapzO2dQAvUptlFV0QLdWoPBMuPI6L5xRXOxVV9a8OwJ0UnfdG6uF0ZqQ10u+zZ/TrEpXnPt66ieL7Z+rpsVCJkkTOxcfDUxGlwaqgvu0smSyowRf9e0tAL3aXMk+zNU4pyWcMl+5W+flDF/G95JQviwdeetN/H/vLRv3ZLk/s5dQdO+9lY7pLoeayFYFcE9DCOnZWYTwT+JUpF34rtJhD8zInIkwY0XbJlaleIrNF4Cb/dEaafjyDcOU60BdmnqeYbSSXhbqj9Rguu7NW+TcTlvwPUfEjazdQpboUu8ux66KU8IuFWL1VxCNTmgxWIRVMNFzfYdvnA3cnvPCdN4Dp+ObdoKz9uwNRGBJPWRiPmWhS30tMuwAm0IebdID3uIZOiCHFNu1NzLB/6cpih1o/oHvHSaamXg0aeCrHu63UTb29tmURXvyiitu0hdbV3z89ee+U+iWXd7UxHwLUOPXjGw17UxiAcbSfCjRVQNZqv1+T8=), along with determining the largest error by projecting along the normals from the original curve, and logging plus visualization of the found largest error.

Note that this does not perform any subdivision yet, so you will naturally find large errors. But this is yielding some really good results. I will now merge this method into the code that @iconexperience has already been working on based on a different method for the offsetting part.

**Comment by lehni** [comment]
I have some good news. The code is working rather beautifully in my local tests. I will put something online for people to experiment more soon, but here for now a screenshot.

<img width="730" alt="screen shot 2017-02-05 at 15 45 10" src="https://cloud.githubusercontent.com/assets/213275/22627019/417b70d0-ebba-11e6-9c63-da122a15d008.png">

**Comment by iconexperience** [comment]
That is quite a useful agorithm. In my original code I could offset through a point at t=0.5, but this one is better, as it allows for arbitrary points. I think using the point where the curve's tangent is the average of the start and end tangent could give better results than simply splitting at t=0.5.

Here is the code that I use to find points on a curve that have a certain tangent:
```javascript
 /**
   * Returns the t values where the curve has a tangent in the same direction as the specified vector.
   */
var getTangentTs = function (v, vTan) {
    var ax = -v[0] + 3 * v[2] - 3 * v[4] + v[6],
      bx = 3 * v[0] - 6 * v[2] + 3 * v[4],
      cx = -3 * v[0] + 3 * v[2],
      ay = -v[1] + 3 * v[3] - 3 * v[5] + v[7],
      by = 3 * v[1] - 6 * v[3] + 3 * v[5],
      cy = -3 * v[1] + 3 * v[3],
      roots = [],
      sTan;
    if (vTan.y < vTan.x) {
      sTan = vTan.y / vTan.x;
      paper.Numerical.solveQuadratic(3 * (ay - sTan * ax), 2 * (by - sTan * bx), cy - sTan * cx, roots, 0, 1);
    } else {
      sTan = vTan.x / vTan.y;
      paper.Numerical.solveQuadratic(3 * (ax - sTan * ay), 2 * (bx - sTan * by), cx - sTan * cy, roots, 0, 1);
    }
    return roots;
  };
```

**Comment by lehni** [comment]
@iconexperience I have been comparing to your original code also, and I believe that this algorithm produces the same shapes with t = 0.5! I am not 100% certain yet, but will soon put a little test-case together that allows us to easily compare these approaches. The code I linked to above also only works with t = 0.5, but based on your input I have now made a more general version of it that can work with any values of t, and will now try your suggestion. I'll put all my tests online soon, but things are behaving surprisingly well already. The trick really is to determine the maximum error and keep subdividing if it is too large.

**Comment by iconexperience** [comment]
Even more important than the maximum error is to check if one of the the handles' direction flipped during offset. If this is the case, it's is a clear sign that the result is not only unsatisfying, but quite messed up. So first check for flipping handles, then iterate until the maximum error is below you threshold.

**Comment by lehni** [comment]
That's a good point too! I've implemented their more general shape control method now that works with an arbitrary curve-time value, and was using the curve-time that has the average tangent for it, but this is producing results not as good as when using t = 0.5. But you are right, when splitting, using this "average tangent curve-time" is much better!

**Comment by lehni** [comment]
[Here a useful sketch](http://sketch.paperjs.org/#S/nVhtc9o4EP4rOr5gGnBsYwgvzd306Id2pi+ZSeZubggfjC1ArbBcWaahmfzg+xe3K78gG9PpHBOwpV3trvblWSnPnTjY086sc/+VqnDX6XdCEeH4+pq8SRIpgnBHxIb8+e8PRiUJM3mgMN6kVLF4S9ZBSiMi4tljfH2NX0LIY+cN+US/k/tdkFCyELGSgpMgjsiCB2nKNiwMFBMx2QhJFtmahaX0BUpPHzuFoPsdG3xiMSX/BKAKBXwEnYMPDIfvMvjNtT7Gh0AWRpFbMnTm+QyVUsiHnaTpTvAIKI7tjebIT4rPcuk6Tp94jrPqkzjjvE/ymQH8rlZ9k9M/cS4HBlOp/xDwjKagpZJZ/OjhsHqr5Mz1NutLfc+xXcefjseTiT/y3RsfeEcTe+rcjEfjqTO+8Yc457sT4HMn8DedDL2bIbJ5tnsznThT1/P9iTf1kM2DpaPRyBm7I9+ZojR/bPuON76ZeCB/NL5Bi0b2dDoFMTe+7w9xoed7tuODuuF0BLZO/dW83GcSqB2YGkoaKHoHAys3v6c5cEeMfrd/CLEHLlfHAidCGisI8a1ebyciZZgDldQ8sQrqhslU6WSY16N7l+vWTCEXMbV6Nkxbz3mgthmL6IwomdEidBuI6fGechoqGuWUx/ilNzeFLgrVJxVNAzZZHOqMPd80KVSbe4gh+fViwzPII6nKZKzJWsTzKb1Sut2Dg9IZWWopdjHh9klt7Jk5mUJhfaULwYWcke6aB+HXrkFu2fvZ4vsw4FBTM7IJeEpzqnbPS33feRHnm9IG9UkiWKxq+09czAttria6hrrEq9M8g6ZO67ZUPUBdw07fqAe2p5bTMxm9y4yuyRhpIFCuHUqRppbyeuQVGZJr4ptMmEloi73PuGIJP1oepFO2VjIIlZW4dhBFVuL1TMkBLIoMsdeoy6CvTbpb0Iv4l5FU7wDLOHWB4aQ66PXa2TzQc2IbrHvn4dkWSfwBwPI9lllKNSG1DiEE6gm+xz45wPNwrCIGlboIeJhxyGgA1y2nRAmidpQ8DYInlhILks/p2af45ky35COWSKCC2LMGBy3Y9FDK4pIHXi29yKSHIi3p8HpOB7ssyZ9AueRH8OAtsRw0pMZxshwNzguvQFFoVjgnhQJyVHjztFgeQgRas46kEEqj76qIADQmCzfMsGfM4fGaTPBxdUu8yn+lV56A6RAu2YoM0NUnIn6OBZFcEVczHOc1U+wkS3dWfc0T5Co6aQCrX6E7++d0dPKVpgNnmTgvVWDvBQeHMKVb7FF3PvI3BeQBh0CQWRzyLALfIQXaku6tHGoGEv6DyLuz1aukRXlabBiH1AK06KMM4IKTAdsnQkImKEKjLQQCjgNp4exFnsZoiW7ylsRkxJzq5x7Xb24DHDWlTHAjxbMkgnBalfd3ujbuzF5iYsMd1nUJIfbovGQkDcsUOgltg7kWRaXJRvsoAK+BfxCiyqJPQu4DXqEaRC5f3iIrV/lTULxkQRNlL1rg/oIFKGxwEW7nLT4zZPQvB+OyW5BYmlUq0J1VZDEm69jE70PVsxeltL80BFgmnOhDoK7jfE6PdTgPLGVrDWi6+Z1qn1TF7+bFr9XD69XVWfWjVQxQXrM0CjW5lJGq1+Cs5W7NJWesJVj9BPMB7m3EfBuwOca32D42xUQsRdO9ZhbUXPAlxwcOD63W5jTeqt0cCK8Jh0fDIaVTvp1Hxty+lrX8smrapO3CpowL3oKBQRxS61tvXmdjG2JFYABuodciId9ZZKx6qccMUqDqYuu0SFqAWy3QWIaKkPf3xk3ibM/lic7GYFjP5zZtpNjPSNKyXyVgHgoirgLRwlQ76P1mhoL8Qbr7AOsy6JJZvvfX5VUIaEJi1SKpK2nUvSi77RxYea9nS7oXB/o5bttbJIPt2emyIorvcXnsrstsD0/D5Y3wlrUMz3rLAxCCRkNtLrZWvrAEfIz3CcKLw7mO1YLJkFfR0nAJp288apRnAxlELIMj+ai8TDDOy9P2VlIal/5s9+FLr7q9VKDz/wyAO2FNU2mEEdICzArlv2rXjqnqXQnBC/Me4NXKuXDWFvFHkaX0LQQT0bLooRY9UOMGALKwdqX4AkBkw+iBpirnyZtR3ywbBeGSWN8zuCOap8fqItTIqDyI5/MYFDPDTreXmumQpJdNx6wDg2t1XczZTFG4yN7eGll0Vv9thxE4K+Z7jyhXwZyQBoyb5w+zFgiFILWBKv6fBrbQUoHdwmldSAnt6bY67+ZGDliMbPngffwzTpGpE+tnGDRKeInuUceErloAGq0985P2FSwq7F0i06rpqYaW+rA8Bs5/0ZsVSpwyAu527FAyGvKaMjr9zhqu+1+1R9PObLl6+Q8=) that visualizes the encountered errors:

* Orange = offset is too close
* Red = offset is too far
* magenta = offset is missing

**Comment by lehni** [comment]
And here now the status quo of the process work:

https://bl.ocks.org/lehni/raw/a665d6f9d95dd055b0ff901f8e313780/

**Comment by lehni** [comment]
Despite the silence here I've been hard at work at getting the offsetting code work reliably. Things have far progressed, to the point where most edge-cases seem squashed, self-intersections are dealt with, and `strokeJoin` is already supported. Here a screenshot:

<img width="662" alt="screen shot 2017-02-22 at 21 38 34" src="https://cloud.githubusercontent.com/assets/213275/23231270/7e34688e-f947-11e6-9dc5-970bcab8293d.png">

**Comment by lehni** [comment]
I am aware that this issue mainly mentions parametrized / variable width strokes. I first want to get fixed width strokes to work well, and then tackle that next. I do have some ideas as to how this algorithm can be extended to support variable width strokes.

**Comment by microbians** [comment]
Impressive work @lehni

**Comment by lehni** [comment]
Here the latest state of the offsetting code, now with boolean operations in the mix:

http://bl.ocks.org/lehni/raw/9aa7d593235f04a3915ac4cef92def02/

Unfortunately, many edge cases remain. I will compile a selection of tricky paths and give access through a pulldown menu, but please have a go and let me know your thoughts and observations.

**Comment by davelab6** [comment]
Loving it!

**Comment by BrianHanechak** [comment]
This already seems like it would be really useful in an application I'm working on, even if there are still edge cases that don't work. Is there any chance that this could be merged in soon?

**Comment by lehni** [comment]
I'm sitting on the fence about that, because it does have a lot of edge cases related to boolean operations still, and endless topic for us unfortunately... But yes, it would be great to add it soon.

**Comment by BrianHanechak** [comment]
I've tested some of the code from your demo, and it looks like it will work really well for my application. 

There's a slight issue with closed paths and the joinOffsets function... basically the resulting path misses a chunk around the starting point of the input path. I can work around this by "unclosing" my path (making a new open path with an extra copy of the first point at the end.)

**Comment by lehni** [comment]
@BrianHanechak can you give an example of that problem?

**Comment by BrianHanechak** [comment]
I'm working on it but having trouble isolating it. (I'm running this inside of a much larger project). Since I'm having trouble isolating it, I'm beginning to think the problem is being caused by other things in my code.

I do get a similar issue if I pass in a compound path, which I am able to show in a fork of your gist: 

https://bl.ocks.org/BrianHanechak/raw/98b0a80656c11b2a7a1a6026932298f2/

In this case, I assume it's just that compound paths aren't supported yet. But the result is similar to what I'm seeing in my bug, so I figured I'd share. I'll keep trying to track this down.

**Comment by sapics** [comment]
@lehni It's really great work!

I have a suggestion about parameter setting.
Currently, parameter (like strokeJoin) comes from the path style parameter.
I would like to set it by option (like below example), when you publish the function.
```javascript
offsetPath: function(path, options){
	if(!options) options = {};
	if(options.offset == null)
		options.offset = path.getStrokeWidth() / 2;
	if(options.strokeJoin == null)
		options.strokeJoin = path.getStrokeJoin();
	if(options.miterLimit == null)
		options.miterLimit = path.getMiterLimit();
	if(options.strokeCap == null)
		options.strokeCap = path.getStrokeCap();
	...
},
```

**Comment by sapics** [comment]
I found one case which looks losing some handles.
It is possible to replicate by console command.
```javascript
path.pathData = "M274.01539,113.5838c-183.54286,28.92795 -238.08075,169.76149 -70.49814,239.58758c167.58261,69.82609 78.53292,249.69441 168.30932,205.80373";
$('#slider-offset').trigger('input');
```

**Comment by lehni** [comment]
Thanks @sapics! Yes the suggestion to optionally provide overrides for these values makes sense.

And wow, that's a funny case you found there. I am aware of a few as well, check these out:

```js
path.pathData = "M466,467c0,0 -105,-235 0,0c-376.816,-119.63846 -469.06596,-146.09389 -650.61329,-266.59735c-282.68388,-230.49081 300.86045,-10.26825 452.77726,121.52815z";
path.pathData = "M466,467c-65,-34 136,64 0,0c-391,-270 62,-670 62,-670l-463,370z";
path.pathData = "M466,467c-65,-34 136,64 0,0c-391,-270 520,-471 522,-137c-214,-144 -1489,123 -923,-163z";
```

The 2nd two are linked to boolean operation issues still... 

So it'll still take quite a bit of time unfortunately to get this all ready for prime-time. The question is, do we included it before?

**Comment by sapics** [comment]
These 3 cases have the similar special situations that neighbor segment.point has same position.
I am happy to use this function before fix these special cases, because it looks rare case in real use case and it might take a lot of time to fix it.
Personally, it is possible to publish or put this function to `develop` branch, because this function have enough a practical use.
It's also possible to make a new branch (like `offset-path`) before publishing, it will make more reports from the participants and watchers.

```javascript
path.pathData = "M274.01539,113.5838c-183.54286,28.92795 -235.08075,170.76149 -67.49814,240.58758c167.58261,69.82609 239.53292,-6.30559 329.30932,-50.19627"
```
This case might be without boolean operations.

**Comment by sapics** [comment]
@lehni I try to fix the case which looks losing some handles where I introduced [before](https://github.com/paperjs/paper.js/issues/371#issuecomment-295546647).
http://bl.ocks.org/sapics/raw/c39df4e0643a404ca90619db63444d5a/

I cannot judge that this fix is good or bad, but, the result looks better.
You can check the difference in https://gist.github.com/sapics/d10f455b8ec1e54a46cf04ed2b386334/revisions

**Comment by microbians** [comment]
Sorry for the semi-offtopic but added variable offsetting to my shit http://microbians.com/?page=code&id=code-bezieroffsetingplayground ;)

**Comment by lehni** [comment]
@microbians not off topic :) Looks great! Unfortunately I haven't had any time lately to work on anything paper.js related, and that won't change for quite a while still : /

**Comment by melchiar** [comment]
Great work! This feature would be very useful in a project that I'm currently working on. Specifically, I've been looking for a way to erode/bloat a complex path. Has work on this stalled for now, or is there any chance of it being merged in the relatively near future? :)

**Comment by luz-alphacode** [comment]
It seems this feature is still missing officially. I was design an animation editing tool and ran into this problem. I found that by the offset algorithm, there are some unwanted self-intersections. I was intend to implement same funtionalities as Adobe Illustrator's offset path and expand stroke.
I managed to cover most of the cases, and I write it as a small extend library for paper.js. I just want to share it here in case someone needs it.
The pacakge name is [paperjs-offset](https://github.com/luz-alphacode/paperjs-offset). It append two methods for Path & CompoundPath: **offset** and **offsetStroke**. Here is the running result of the lib.
![preview](https://raw.githubusercontent.com/luz-alphacode/paperjs-offset/master/public/preview.jpg)

**Comment by lehni** [comment]
@luz-alphacode this looks great! Did you see my version posted above at http://bl.ocks.org/lehni/raw/9aa7d593235f04a3915ac4cef92def02/ ?

The actual code you can see here: http://bl.ocks.org/lehni/raw/9aa7d593235f04a3915ac4cef92def02/offset.js
This includes code that handles joins and ends, through Paper's own already existing internal functions, such as `Path._addBevelJoin()`, `Path._addSquareCap()`, etc.

The offsetting code itself works really well and is based on  __A New Shape Control and Classification for Cubic Bezier Curves__ by Shi-Nine Yang and Ming-Liang Huang.

The only reason why I haven't rolled this into the library itself is issues with self intersection when expanding the offset to outlines. These issues are in the boolean operations code, and I so far have not managed to find more time to try and resolve them : (

**Comment by luz-alphacode** [comment]
Glad for your invitation, I was a bit busy these days, I'd like to check this few days later. @lehni

**Comment by melchiar** [comment]
Great to see progress on this feature! If it's of any help, I've run a few tests using the latest code from @lehni and found a bit of weirdness in certain cases where a point's handles are aligned perfectly (example below)

![image](https://user-images.githubusercontent.com/9057412/57003729-7cc6d400-6b7d-11e9-831e-1061d7544662.png)

`var pathData = "M288 216H0V44.514c21.206 0 32.848-.841 62.371-16.358C81.174 18.271 107.746 0 143.999 0c36.255 0 62.827 18.271 81.629 28.155C255.15 43.673 266.794 44.514 288 44.514V216z";`

**Comment by shagarah** [comment]
Any updates / release times please?

**Comment by lehni** [comment]
@shagarah I've worked some more on it recently, but am again swamped with work currently, so can't give a timeline. I want to release this soon as part of a final v1.0.0, but it needs more tweaking to be reliable enough.

**Comment by northamerican** [comment]
@lehni i'd love to see an updated gist with the recent changes. thanks.

one great thing with @luz-alphacode 's is that it removes 'loops' like these:

<img width="85" alt="Screen Shot 2019-08-05 at 2 22 53 PM" src="https://user-images.githubusercontent.com/6657634/62486178-89feea80-b78c-11e9-8fb5-bfc819d8e1c3.png">

dunno if you have an effective way of removing those? i just check the distance of curves and remove them if they are too 'close' to the original path, but it's not efficient.

**Comment by hujiulong** [comment]
Hey guys, is there any progress on this?

**Comment by northamerican** [comment]
i can offer a bounty to anyone able to iron out the remaining issues with the offset code.

**Comment by northamerican** [comment]
i can offer a financial bounty to anyone able to iron out the remaining issues with the offset code.

**Comment by thexperiments** [comment]
Wow I did not realize how complicated that is. Started some project with paper.js assuming this would be a nobrainer and now I‘m stuck. 

What is the current status here? 

Are there any workarounds via external libraries?

Thanks

**Comment by lehni** [comment]
@northamerican I should have time to start working on this again in about 6 weeks. Please get in touch at juerg@lehni.org so we can discuss this further. Thanks!

**Comment by lehni** [comment]
@northamerican did you try to get in touch? I didn't receive anything :)

**Comment by northamerican** [comment]
@lehni e-mail has been sent. ready to discuss when you are.

**Comment by martin19** [comment]
@lehni: Thanks very much for your work on this. I've got a question/comment about your offset code posted above: If I understand correctly the connect method works by connecting offset segments using the join-type set through the paths stroke style (round, bevel).  Is there an exact definition on how the outline feature should work in this regard ? Is there a way to connect segments by "elongating" them and computing their intersections - this works for straight lines - I'm not sure about bezier curves? About your statement above: "These issues are in the boolean operations code" - can you give some more hints on what is going wrong? I haven't found a reference to the boolean op code in your snippet. Thanks!

**Comment by NoZ4** [comment]
I have written code for paper.js for offsetting, please find it here : https://github.com/NoZ4/paper.js-offsets
Im not good with javascript or mathamatics and its taken me 6 months to produce this. Its for procedual generation of cad designs for laser cutting

I found offsetting curves and paths are easy, its the self intersections that are the real problem along with floating point errors
so in my code I work out where self intersections will occur first and trim curves before offsetting them

Also even if curves dissapper becouse of offsetting the points where curves originally touched must be taken into account I have code creating paths along these points, with arcs.

it works with compound paths, using winding rules, so doughnut shapes can be shrunk/expandid.

I know my code has better accuracy than inkscape, i haven't tested against others. 

Im quite cheery that if you expand enough the paths become a circle, or if shrunk turns into a point

it takes into account where paths totally disapper becouse they cant be offset that much and curves/paths inverseing

I got examples on github that show what i mean.

I really hope this can be of use otherwise ive wasted alot of time on nothing

**Comment by michaeltford** [comment]
Hi, first I wanted to say that I recently started using paper.js and I am amazed. I am porting a tool that renders shapes from Java to Javascript and paper.js is making this very easy. There is one problem that I am am not able to figure out. 

I need to be able to render complex paths and find the bounds of these paths. It looks like I can use offset (if it's ready?) but is there a way to render double strokes or other styles? (Similar to below?).

http://www.jhlabs.com/java/java2d/strokes/

My most basic use case is to draw and capture the bounding rect of a double/triple dashed line (Similar to what PowerPoint calls a Compound Line) 

https://www.officetohelp.com/powerpoint/how-to-make-compound-line-in-ppt.html

In Java2D these can be done by create a 'Brush' that walks path segments and allows for the ability to render something at each segment. (The work involved is very complicated because of rotations, miter sizes, drawing joins and all the other complicated things that people here understand better than me.

**Comment by NoZ4** [comment]
hiya,  
with proper working offsets double strokes should be really easy, 5 offsets would be needid. so you have your path which would be the center of your line, you offset it + and - then you have two new lines you just need to cap off, then if you want a double line you offset the original path + or - , just more than you did before and repeat what you did for the original

dashed line should be easy just create gaps in the inital path.

brushing is something we dont have at the moment i think, but shouldnt be too hard to impliment, get even points along a path and just clone whatever you want onto that posistion.

Im currently working on my code, Im not doing double/triple line but am doing lines with variable thickness etc i'll be updating in about a week if thats any use

**Comment by NoZ4** [comment]
Hiya everyone

Just thought i would give a update, my code now allows to stroke a path
[https://github.com/NoZ4/paper.js-offsets/blob/master/VariableWidthstroke.SVG](url)
[https://github.com/NoZ4/paper.js-offsets/blob/master/Stroke.SVG](url)
[https://github.com/NoZ4/paper.js-offsets/blob/master/InverseStroke.SVG](url)

theres a few known issues, shown on the page

@lehni do you think it would ever be a idea to add this to the main code? also I used a bit of your code for offsetting single curves

I think this might be a bit quicker than lehni's approach, since i got some escape early functions, but for things like square caps etc I think lehni's approach is much better. 

Thanks for all the thumbs up etc before :) 

Thanks everyone  Noz

**Comment by northamerican** [comment]
hey, seekers of path offsets. 
i have yet another option, [paper-clipper](https://github.com/northamerican/paper-clipper) to share with you. as @lehni wrote above in 2014, the Clipper library handles polygons only. however, it is very accurate in creating offsets for paths without segment handles. this library harnesses Clipper's performant boolean operations (see clipperUnite) as well as its offset functionality. it flattens Paper Paths, offsets them then applies a specialized `simplify` to restore curves, with good results.

with this external dependency, it isn't viable for inclusion in Paper of course but it's performant and reliable enough for my artistic uses. maybe it will come in use until we have native functionality working and approved.

if you are curious as to how it applies smoothing to the path after offsetting, refer to [paper-clipper/betterSimplify.ts](https://github.com/northamerican/paper-clipper/blob/master/src/betterSimplify.ts). it uses Paper's [`Path#simplify`](http://paperjs.org/reference/path/#simplify) method on split up pieces of the resulting offset path. this helps retain sharper corners which `simplify` tends to overcompensate for. 

![strands-export-2020-09-11T03_22-01](https://user-images.githubusercontent.com/6657634/92853138-79706900-f3bd-11ea-84c1-90bfe352d78c.png)

**Comment by NoZ4** [comment]
Hello, I thought since this is still a active thing i'll add some bits of info to help anyone working on a solution like im trying to

So here is a undivided curve, curves in solid color, handles in dots
red = initial curve
green = scaling curve (from @lehni code)
yellow = moving original handles by each handles normal and then reconstructing handles from new intersects


![curve_test1](https://user-images.githubusercontent.com/61328968/118861714-b7227b80-b8d4-11eb-9eac-f7bb74cc0a91.png)

![curve_test](https://user-images.githubusercontent.com/61328968/118861749-c1dd1080-b8d4-11eb-9744-d89bf67554f1.png)

I tried using average and middle offsets from @lehni's code but didnt seem to work ( Im thinking they probably have to be more resonable curves)

Also I know we break curves up to get better accuracy but this is just a little guide for peeps before that
